### PR TITLE
Adds Matching resource

### PR DESCRIPTION
**Pull request by OmarDarwish**
This resource lets you define arbitrary data to match and report against. I use this when piping in a generated file and using templates to match against its contents:

Let's say we have a `data.json` file that gets generated as part of some testing pipeline:

```json
{
  "instance_count": 14,
  "failures": 3,
  "status": "FAIL"
}
```

This could be piped into goss: `goss --vars data.json validate`

And then validated against:

```yaml
---
matching:
  check_instance_count: # Make sure there is at least one instance
    content: {{ .Vars.instance_count }}
    matches:
      gt: 0

  check_failure_count_from_all_instance: # expect no failures
    content: {{ .Vars.failure_count }}
    matches: 0

  check_status:
    content: {{ .Vars.status }}
    matches:
      - not: FAIL
```

### CODE REVIEW

**Comment by OmarDarwish** [comment]
This PR should be merged after #225 and #233

**Comment by OmarDarwish** [comment]
Once #256 is merged, the tests for this should be passing

**Comment by OmarDarwish** [comment]
@aelsabbahy do you think this will make it into the next release?

File: docs/manual.md [hunk 1]
```diff
@@ -614,6 +615,30 @@ mount:
     filesystem: xfs
 ```
 
+### matching
```
Comment (aelsabbahy) [inline docs/manual.md]: Can you provide an example with templates for the readme? or a brief explanation like you did on this PR.

File: resource/matching.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	stub := func() (interface{}, error) {
+		return a.Content, nil
+	}
+
+	var results []TestResult
+	results = append(results, ValidateValue(a, "matches", a.Matches, stub, skip))
+	return results
+}
+
+func (ret *MatchingMap) UnmarshalJSON(data []byte) error {
```
Comment (aelsabbahy) [inline resource/matching.go]: Can this be moved to resource_list_genny.go and generated into resource_list.go? Or is there something special about this particular implementation?

Comment (OmarDarwish) [inline resource/matching.go]: I think this resource is unique in that it does not have a system resource pairing. The template in resource_genny attempts to instantiate a matching system resource. I can add a dummy resource if you think that approach is better.

File: docs/manual.md (hunk 1 above)
Comment (OmarDarwish) [inline docs/manual.md]: Yes, can do.

File: resource/matching.go (hunk 2 above)
Comment (aelsabbahy) [inline resource/matching.go]: We can leave it as is for now. If a second non-system resource comes up, then the dummy approach will be preferable.

File: docs/manual.md (hunk 1 above)
Comment (OmarDarwish) [inline docs/manual.md]: Done.

**Comment by aelsabbahy** [comment]
Was doing some thinking about this PR last night. Some high level thoughts on a V2 implementation of this.

What if this was a validator type, that can be used on any JASON/YAML file or url pointing to one. An optional attribute for both the http and file resource. The tricky part would be the template portion, or maybe we can see if go provides a library that does something like jq.

This implementation works, but feels limited and awkward to use the vars file for this.

Any thoughts?

**Comment by elliotweiser** [comment]
@aelsabbahy (in response to your question...)

IMO, this starts to feel like feature-creep. I like the idea of JSON/YAML file validation, but I think that work should be externalized to a different tool/library that could then be called from GOSS. I know of tools that already do this e.g. [yaml_bot](skippyPeanutButter/yaml_bot). Incidentally, I think @skippyPeanutButter has started work on a Golang version of that tool ☝️(and is looking for collaborators). 

Personally, I would like to see this functionality moved to a validator for the `file` resource. GOSS could then call custom rules validation engines then be be called to validate its structure/contents.  Here's a rough example in my head...

```yaml
file:
  /path/to/my.json
    exists: true
    filetype: file
    validator:
      jsonValidatorLibrary:
        with_tests: /path/to/json/file/tests
```

To summarize, I think it's a good idea, but it feels slightly out-of-scope and should probably be handled in a separate library/project. 

That is my two cents, but I may be in the minority here.
